airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/DestinationTaskExceptionHandler.kt

```diff
@@ -0,0 +1,148 @@
+    private val syncManager: SyncManager,
+    private val failStreamTaskFactory: FailStreamTaskFactory,
+    private val failSyncTaskFactory: FailSyncTaskFactory,
+) : DestinationTaskExceptionHandler<ScopedTask> {
+    val log = KotlinLogging.logger {}
+
+    inner class SyncTaskWrapper(
+        private val syncManager: SyncManager,
+        private val innerTask: SyncTask,
+    ) : SyncTask, InternalTask {
+        override suspend fun execute() {
+            if (!syncManager.isActive()) {
+                val result = syncManager.awaitSyncResult()
+                if (result is SyncSuccess) {
+                    throw IllegalStateException(
+                        "Task $innerTask run after sync has succeeded. This should not happen."
+                    )
+                }
+                log.info { "Sync terminated, skipping task $innerTask." }
+
+                return
+            }
+
+            try {
+                innerTask.execute()
+            } catch (e: Exception) {
+                handleSyncFailure(e)
+            }
+        }
+
+        override fun toString(): String {
+            return "SyncTaskWrapper(innerTask=$innerTask)"
+        }
+    }
+
+    inner class StreamTaskWrapper(
+        private val syncManager: SyncManager,
+        private val innerTask: StreamTask,
+    ) : SyncTask, InternalTask {
+        override suspend fun execute() {
+            // Stop dispatching tasks if the stream has been killed by a failure elsewhere.
+            // Specifically fail if the stream was marked succeeded: we should not be in this state.
+            val streamManager = syncManager.getStreamManager(innerTask.stream.descriptor)
+            if (!streamManager.isActive()) {
+                val result = streamManager.awaitStreamResult()
+                if (result is StreamSucceeded) {
+                    throw IllegalStateException(
+                        "Task $innerTask run after its stream ${innerTask.stream.descriptor} has succeeded. This should not happen."
+                    )
+                }
+                log.info {
+                    "Stream ${innerTask.stream.descriptor} terminated with $result, skipping task $innerTask."
+                }
+                return
+            }
+
+            try {
+                innerTask.execute()
+            } catch (e: Exception) {
+                handleStreamFailure(innerTask.stream, e)
+            }
+        }
+
+        override fun toString(): String {
+            return "StreamTaskWrapper(innerTask=$innerTask)"
+        }
+    }
+
+    override fun withExceptionHandling(task: LeveledTask): ScopedTask {
+        return when (task) {
+            is SyncTask -> SyncTaskWrapper(syncManager, task)
+            is StreamTask -> SyncTaskWrapper(syncManager, StreamTaskWrapper(syncManager, task))
+        }
+    }
+
+    override suspend fun handleSyncFailure(e: Exception) {
+        log.error { "Sync failed: $e: killing remaining streams" }
```

Comment:
nit: (preserve the stacktrace)
```suggestion
        log.error(e) { "Sync failed: killing remaining streams" }
```

Reply:
This will get logged at the end because it will go into the sync failure.